Index: src/cmd/devdraw/cocoa-screen.m
===================================================================
--- a/src/cmd/devdraw/cocoa-screen.m
+++ b/src/cmd/devdraw/cocoa-screen.m
@@ -742,9 +742,13 @@
 			if(m & NSAlternateKeyMask){
 				abortcompose();
 				b = 2;
-			}else
-			if(m & NSCommandKeyMask)
+			}else if(m & NSCommandKeyMask)
 				b = 4;
+			else if(m & NSControlKeyMask){
+				in.mbuttons = 2;
+				sendmouse();
+				b = 1;
+			}
 		}
 		in.mbuttons = b;
 		break;
